### PR TITLE
Make CPU and GPU RNN broadcasting behavior consistent

### LINKED ISSUE #3789: broadcasting behaves differently on CPU and GPU
GRUCell and LSTMCell on GPU do not support broadcasting, while RNNCell supports broadcasting on both CPU and GPU. This inconsistency is not expected and not documented.

Here's the example code:
```python
import torch
import torch.nn as nn
from torch.autograd import Variable

rnn = nn.RNNCell(10, 20)
gru = nn.GRUCell(10, 20)
input = Variable(torch.randn(3, 10))
h0 = Variable(torch.randn(1, 20)) # this should be (3, 20)
# both cells support broadcasting on CPU
hx_rnn = rnn(input, h0)
hx_gru = gru(input, h0)
assert hx_rnn.size() == hx_gru.size()

# send models and variables to GPU
rnn.cuda()
gru.cuda()
input = input.cuda()
h0 = h0.cuda()

# RNNCell is OK
try:
    hx_rnn_cuda = rnn(input, h0)
except RuntimeError as e:
    print('rnn error\n', e)

# GRUCell on GPU does not support broadcasting
try:
    hx_gru_cuda = gru(input, h0)
except RuntimeError as e:
    print('gru error\n', e)
```
The output of this example is:
```
gru error
invalid argument 3: Input and Hidden tensor sizes should be the same. at /pytorch/torch/lib/THCUNN/generic/FusedRNNKernel.cu:19
```

Is this a known issue or not expected?

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #3789. 

Before this change the following behavior occurs:
- RNNCell supports input/hidden broadcasting on CPU and GPU
- GRUCell and LSTMCell support input/hidden broadcasting on CPU but not GPU

This makes it so that GRUCell and LSTMCell support input/hidden broadcasting

### Test Play
Unit testing

### CODE REVIEW

File: torch/nn/_functions/rnn.py [hunk 1]
```diff
@@ -21,8 +21,21 @@ def RNNTanhCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
     return hy
 
 
+def _match_hidden_with_input(input, hidden):
```
Comment (fmassa) [inline torch/nn/_functions/rnn.py]: I wonder if this is similar to `torch._C._infer_size` (which is btw used by [`variable_expandable`](https://github.com/pytorch/pytorch/blob/709fcfda8a2fb3f3da33cace584e6d5592210a9c/torch/autograd/_functions/utils.py#L31-L38) )?

Comment (zou3519) [reply torch/nn/_functions/rnn.py]: `_infer_size(x,y)` looks like it returns the size of broadcasting with x and y. Here I'm only trying to make dim 0 of input and hidden match (dim 1 is something else that doesn't need to match).

Comment (fmassa) [reply torch/nn/_functions/rnn.py]: I see. I was thinking if it is possible to cut-off the dimensions that you don't want to broadcast so that you can still use `_infer_size`, as was done in a [previous version of `matmul`](https://github.com/pytorch/pytorch/pull/1780/files#diff-61b893a141a037c275f12eba9a9b864fR185)

Comment (fmassa) [reply torch/nn/_functions/rnn.py]: But this is a completely minor comment on my side :)

File: torch/nn/_functions/rnn.py [hunk 2]
```diff
@@ -21,8 +21,21 @@ def RNNTanhCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
...
+    '''The CUDNN backend doesn't support broadcasting, so this function
+    broadcasts the hidden batch size to match the input batch size, if
+    possible'''
+    if isinstance(hidden, tuple):
```
Comment (ngimel) [inline torch/nn/_functions/rnn.py]: This comment is misleading, it's not cudnn backend, it is fused LSTM and GRU kernels in pytorch/THCUNN.

Comment (zou3519) [reply torch/nn/_functions/rnn.py]: Thanks for pointing that out, will update!

**Comment by csarofeen** [comment]
I'm worried this is simply enabling error prone user code. If you're going to go through the trouble of sending your own hidden state to RNN instead of leaving it out and letting RNN module handle hidden state initialization to Zero, you should be expected to size it correctly. Maybe the better fix is to throw an error on CPU side.

User reporting issue did not provide a convincing reason IMO as to why this is necessary.

**Comment by apaszke** [comment]
I also think that being more strict in case of RNN cells is a good thing. Not all operations have to support broadcasting.

**Comment by zou3519** [comment]
That sounds reasonable. Closing this; will put up a new PR for stricter error checking in RNNs.
